### ISSUE TITLE
Having Concurrent Later Overlapping Events Hides Earier Event

### ISSUE THREAD

**Issue by Christopher-Chianelli** [intljusticemission/react-big-calendar#1397]
#### Do you want to request a _feature_ or report a _bug_?

_Bug_

#### What's the current behavior?

A regression/edge case of #404 (which should be reopened)

![Image Of Issue](https://user-images.githubusercontent.com/15825810/61739328-376c0a00-ad5a-11e9-93bd-77b4be0622fc.png)

Code Sandbox: https://codesandbox.io/embed/react-big-calendar-example-96cue

This edge case occurs precisely when the following two conditions are satisfied:

1. There is an event A that overlaps with another event B which ends later
2. There are events C and D that begin after A but overlap with B

Affected Browser(s):

- Mozilla Firefox 68.0
- Google Chrome 75.0.3770.142

#### What's the expected behavior?

For event B not to be completely hidden by events C/D. Example (from Google Calendar):
![ReactBigCalendarExpectedResult](https://user-images.githubusercontent.com/15825810/61740463-b06c6100-ad5c-11e9-9d7e-7c0fb23ef6e7.png)

**Comment by dmitrykrylov** [comment]
@Christopher-Chianelli oh, I planned to report a similar bug today.

Here is a codesandbox with a combination of events with same times as I have in a real app.
There are events A - H and event G is **completely** hidden behind other ones.

https://codesandbox.io/s/react-big-calendar-overflowing-events-9hdoh

**Comment by Christopher-Chianelli** [comment]
@dmitrykrylov Depending on how important it is to be able to view overlapping shifts and how important having a Google Calendar-like calendar is to you, there is a workaround that will have all shifts visible: https://codesandbox.io/s/zqk8vmqn3m (from #814). The disadvantages of this approach are:

1. It differs from Google Calendar.
2. Having lots of concurrent events at the same time mean events without any concurrent events are extremely narrow.
3. You need to hack the position and display styles of the components.

I am currently using this approach on an application; it works quite well even with 6 or more concurrent events.

**Comment by dmitrykrylov** [comment]
@Christopher-Chianelli I like this way to display events! But it seems like it doesn't take minutes into account?

**Comment by Christopher-Chianelli** [comment]
@dmitrykrylov What I did was overwrite the wrapper to use style instead of `startDate.hours`. `top` will be between `0%` and `100%`, and `height` will be between `0%` and `100%`. `gridStartPoisition` would than be `Math.floor(parseFloat(style.top) * 0.01 * rowsInGrid))` and `gridEndPosition` will be `Math.floor(parseFloat(style.height) * 0.01 * rowsInGrid) + gridStartPosition`. I warn against using more than 1000 rows in a grid; some browsers will not render those properly: https://bugs.chromium.org/p/chromium/issues/detail?id=688640. As such, I recommend `rowsInGrid = 770` for 2 minute precision, or `rowsInGrid = 100` if you want to simpilfy the formula to `parseInt(style.top)`. (Of course, using the event start & end times won't break if in the future they don't use percent `widths` and `heights`).

**Comment by dmitrykrylov** [comment]
@Christopher-Chianelli awesome! Haven't you managed to make it working with drag'n'drop?

**Comment by Christopher-Chianelli** [comment]
@dmitrykrylov Unfortunately no; if you are able to detect when the drag starts, you might be able to do it by temporarily disabling the modified styles (which can be done by moving the overrides into a new CSS classes, and only adding them to the `EventWrapper` and `Event` if you are not dragging).

**Comment by dmitrykrylov** [comment]
@Christopher-Chianelli I tried to rewrite the events arrangement algorithm

Before
<img width="791" alt="Screen Shot 2019-07-31 at 13 56 42" src="https://user-images.githubusercontent.com/14343880/62206766-34e16400-b39b-11e9-8589-e5e1ff22ba4e.png">

After
<img width="757" alt="Screen Shot 2019-07-31 at 13 56 16" src="https://user-images.githubusercontent.com/14343880/62206777-3ad74500-b39b-11e9-88e0-1377205b04f3.png">

**Comment by Christopher-Chianelli** [comment]
@dmitrykrylov That looks a lot better!

**Comment by WillisShek** [comment]
@dmitrykrylov It looks great!

Besides getting a new algorithm. It's possible to inject a new prop `numberOfConcurrentEvents` to the `EventWrapper` based on the new algorithm? It opens a large door to the customization of the styles of those events, like allowing non-overlapping events or a different background color depending on whether the event is overlapped by the others.

**Comment by Christopher-Chianelli** [comment]
This issue might be fixed by https://github.com/intljusticemission/react-big-calendar/pull/1473; I will verify this issue is fixed when a new version is released.

**Comment by Christopher-Chianelli** [comment]
Can confirm #1473 fixes the issue when the Calendar prop `dayLayoutAlgorithm="no-overlap"` is set.

**Comment by stiveurkel** [comment]
> @Christopher-Chianelli I tried to rewrite the events arrangement algorithm
> 
> Before
> <img alt="Screen Shot 2019-07-31 at 13 56 42" width="791" src="https://user-images.githubusercontent.com/14343880/62206766-34e16400-b39b-11e9-8589-e5e1ff22ba4e.png">
> 
> After
> <img alt="Screen Shot 2019-07-31 at 13 56 16" width="757" src="https://user-images.githubusercontent.com/14343880/62206777-3ad74500-b39b-11e9-88e0-1377205b04f3.png">

@Christopher-Chianelli  Can you please let me know where in the code have you made these amendments? Curious about the logic behind it.

**Comment by Christopher-Chianelli** [comment]
@stiveurkel I believe you pinged the wrong person. @dmitrykrylov made those changes.

**Comment by stiveurkel** [comment]
@dmitrykrylov Hi Dmitry, Christopher mentioned you have made the event overlapping amendments, it's been awesome so far! Would you be able to help with that issue I've raised yesterday?
https://github.com/jquense/react-big-calendar/issues/1839

**Comment by dmitrykrylov** [comment]
Hi @stiveurkel. My PR #1405 was ignored, however, #1473 should fix your issues.

**Comment by stiveurkel** [comment]
@dmitrykrylov Actually it does not, #1473 is causing the overlapping (the screenshot on the issue I've raised). I have implemented your PR (#1405) and it has just fixed it :) Awesome, thanks!

**Comment by dmitrykrylov** [comment]
@stiveurkel hm interesting 😁 Couldn't you fork [this codesandbox](https://codesandbox.io/s/react-big-calendar-example-no-overlap-hx4ob?file=/index.js:566-604) and make a showcase of your issue?

**Comment by stiveurkel** [comment]
Sure, here you go: https://codesandbox.io/s/react-big-calendar-example-no-overlap-forked-t5n3k?file=/events.js

Check the 1 - 5pm on January 3rd 2021, it's hidden behind all other events.

**Comment by dmitrykrylov** [comment]
@stiveurkel oh I see it. Are you applying my updates to the latest version of the library? If yes, maybe you could make your own PR based on my updates?

@ckcks12 hi. You had a big update on the algorithm and overlapping issues. Could you check what's missing in your implementation that doesn't cover the case of @stiveurkel?

**Comment by stiveurkel** [comment]
@dmitrykrylov Yes for my project I am applying your PR (that I slightly tweaked for my needs) which works great :) I have tried with @ckcks12 but it causes some issues every now and then (as you can see on the codepen and on #1839 screenshots).

**Comment by dmitrykrylov** [comment]
@stiveurkel I mean that you can help this project by creating your own PR to react-big-calendar with the changes you made

**Comment by ckcks12** [comment]
![image](https://user-images.githubusercontent.com/12825679/103770415-50a96900-5069-11eb-8f8a-8a526eddf5ed.png)

TL; DR
```
dayLayoutAlgorithm="no-overlap" // overlap -> no-overlap
```


Hi all, Thanks for remember me haha

First of all, I'm disappointed with that maintainer hasn't put my example(demo) into where it should be.
My PR #1473 had an example. Look at [this ](https://github.com/jquense/react-big-calendar/pull/1473/files#diff-ca07cead5fc37873155aae1bede6c849ec54ba268fe2836d088692db35a1148cR43)
I don't know why but it would be much better to include it in official documentation. (way much better)

Hmmm,,,, so sad. 🥲

**Comment by stiveurkel** [comment]
@ckcks12 the point was to have events overlapping, that's why I have put `dayLayoutAlgorithm="overlap"`, it works for most cases, except the ones I have found.

Do you mean that http://jquense.github.io/react-big-calendar/examples/index.html demo example should be no-overlap ?

**Comment by ckcks12** [comment]
Oh actually `overlap` is the basic algorithm. acutally there's no "overlap" algorithm. I named it because the default algorithm is overlapping.

I mean, I created just only the no-overlap algorithm. and make it pluggable ㅡ so let you choose overlap or no-overlap

**Comment by dmitrykrylov** [comment]
In my opinion, the library should have just one algorithm that should arrange events into groups and columns nicely and ideally to be performant. The overlapping feature here shouldn't affect the algorithm and could be implemented just like a boolean prop:
```jsx
<Calendar {...props} overlap />
```
We could take the FullCalendar API as a good example:
https://fullcalendar.io/docs/slotEventOverlap

My PR was just about improving the algorithm, and the overlapping feature could be implemented as a separate contribution.

**Comment by stiveurkel** [comment]
@ckcks12 Ok I see, so the basic algorithm fails sometimes. If you take the dates array from my codepen and compare how it renders on fullcalendar.io vs reactbigcalendar you will see it's 90% similar, few events don't render properly and are hidden.

@dmitrykrylov agree and your PR is doing a great job.

**Comment by Nageswaran-OpsEazy** [comment]
![Image](https://github.com/user-attachments/assets/1cc0a3ec-93fe-4064-9b88-30aa4fc4aa85)

Hi advance thanks for if get solution. 

I have used react-big-calendar in my projects i am getting overlap issue in week view because my event date and time is same.

Example: 
Meeting 1 = 29 Aug 2025, 5pm to 5.30pm.
Meeting 2 = 29 Aug 2025, 5pm to 5.30pm.

Can we add show more button in week view? by default showing in month view but i need to show in week view.
